targets/f7/furi_hal/furi_hal_serial.c

```diff
@@ -0,0 +1,797 @@
+            LL_USART_SetPrescaler(USART1, prescaler);
+            LL_USART_SetBaudRate(USART1, uartclk, prescaler, FURI_HAL_USART_OVERSAMPLING, baud);
+            LL_USART_Enable(USART1);
+        }
+    } else if(handle->id == FuriHalSerialIdLpuart) {
+        if(LL_LPUART_IsEnabled(LPUART1)) {
+            // Wait for transfer complete flag
+            while(!LL_LPUART_IsActiveFlag_TC(LPUART1))
+                ;
+            LL_LPUART_Disable(LPUART1);
+            uint32_t uartclk = LL_RCC_GetLPUARTClockFreq(LL_RCC_LPUART1_CLKSOURCE);
+            LL_LPUART_SetPrescaler(LPUART1, prescaler);
+            LL_LPUART_SetBaudRate(LPUART1, uartclk, prescaler, baud);
+            LL_LPUART_Enable(LPUART1);
+        }
+    }
+}
+
+void furi_hal_serial_deinit(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    furi_hal_serial_async_rx_configure(handle, NULL, NULL);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(furi_hal_bus_is_enabled(FuriHalBusUSART1)) {
+            furi_hal_bus_disable(FuriHalBusUSART1);
+        }
+        if(LL_USART_IsEnabled(USART1)) {
+            LL_USART_Disable(USART1);
+        }
+        furi_hal_serial_usart_deinit_dma_rx();
+        furi_hal_gpio_init(&gpio_usart_tx, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_gpio_init(&gpio_usart_rx, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_serial[handle->id].enabled = false;
+    } else if(handle->id == FuriHalSerialIdLpuart) {
+        if(furi_hal_bus_is_enabled(FuriHalBusLPUART1)) {
+            furi_hal_bus_disable(FuriHalBusLPUART1);
+        }
+        if(LL_LPUART_IsEnabled(LPUART1)) {
+            LL_LPUART_Disable(LPUART1);
+        }
+        furi_hal_serial_lpuart_deinit_dma_rx();
+        furi_hal_gpio_init(&gpio_ext_pc0, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_gpio_init(&gpio_ext_pc1, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_serial[handle->id].enabled = false;
+    }
+}
+
+void furi_hal_serial_suspend(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdLpuart && LL_LPUART_IsEnabled(LPUART1)) {
+        LL_LPUART_Disable(LPUART1);
+    } else if(handle->id == FuriHalSerialIdUsart && LL_USART_IsEnabled(USART1)) {
+        LL_USART_Disable(USART1);
+    }
+    furi_hal_serial[handle->id].enabled = false;
+}
+
+void furi_hal_serial_resume(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    if(!furi_hal_serial[handle->id].enabled) {
+        if(handle->id == FuriHalSerialIdLpuart) {
+            LL_LPUART_Enable(LPUART1);
+        } else if(handle->id == FuriHalSerialIdUsart) {
+            LL_USART_Enable(USART1);
+        }
+        furi_hal_serial[handle->id].enabled = true;
+    }
+}
+
+void furi_hal_serial_tx(FuriHalSerialHandle* handle, const uint8_t* buffer, size_t buffer_size) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(LL_USART_IsEnabled(USART1) == 0) return;
```

Comment:
inconsistent enabled checks - in some places it's LL_*, sometimes [].enabled

Reply:
yes, whole thing must be refactored, but not in this iteration